gerrychain/tree.py

```diff
@@ -151,11 +159,37 @@ def part_nodes(start):
                                 queue.append(c)
             return nodes
 
-        if abs(tree_pop - h.ideal_pop) <= h.ideal_pop * h.epsilon:
-            cuts.append(Cut(edge=(node, pred[node]), subset=part_nodes(node)))
-        elif abs((total_pop - tree_pop) - h.ideal_pop) <= h.ideal_pop * h.epsilon:
-            cuts.append(Cut(edge=(node, pred[node]),
-                            subset=set(h.graph.nodes) - part_nodes(node)))
+        is_balanced_A = abs(tree_pop - h.ideal_pop) <= h.ideal_pop * h.epsilon
+        is_balanced_B = abs((total_pop - tree_pop) - h.ideal_pop) <= h.ideal_pop * h.epsilon
+
+        parent = pred[node]
+        if region_weights is not None:
+            region_cols = [tup[0] for tup in region_weights]
+            # score function to prefer dividing higher ranked regions. three regions -> [4, 2, 1]
+            # an edge that splits our first- and last-ranked regions would get a score of 4 + 1 = 5
+            # if node corresponds to a balance cut and our split score is as good or better than our
+            # best previously seen split score, add this cut to our list of potential balance cuts
+            node_split_score = 0
+            for i, region_col in enumerate(region_cols):
+                if h.graph.nodes[parent][region_col] != h.graph.nodes[node][region_col]:
+                    node_split_score += 2 ** (len(region_cols) - i - 1)
+
+            if node_split_score > best_split_score and (is_balanced_A or is_balanced_B):
```

Comment:
Oddly enough, this might be more clear if it was nested differently to avoid handling the symmetry twice, e.g.
```python
if node_split_score >= best_split_score and (is_balanced_A or is_balanced_B):
  if node_split_score > best_split_score:  # special case: strictly better
    best_split_score = node_split_score
    cuts = []
  part_subset = part_nodes(node) if is_balanced_A \
                    else set(h.graph.nodes) - part_nodes(node)
  cuts.append(Cut(edge=(node, pred[node]), subset=part_subset))
```